### ISSUE TITLE
Consider autoloaders that check for class existence after loading (potentially) multiple files?

### ISSUE THREAD

**Issue by Ocramius** [Roave/BetterReflection#438]
As discussed in #384 (https://github.com/Roave/BetterReflection/issues/384#issuecomment-346900693) and #437, the autoloading mechanism could be like following (pseudo-code):

```php
spl_autoload_register(function (string $className) : bool {
    if (! file_exists(class_to_file($className))) {
        return false;
    }

    require_once class_to_file($className);
    return class_exists($className, false);
});
```

In practice, this means that, given 2 separate files where `require_once` is effectively used, the class may be registered only once.

This is sub-optimal, and isn't currently handled by the `AutoloadSourceLocator`, because `AutoloadSourceLocator` only considers the first included file during autoloading operations. Still, this is an edge case that may need documenting, even if just as this issue.

In general, I don't think that this scenario will occur often enough (if at all) to require fixing, but having an issue that describes it will help anyone hitting this and willing to contribute to the project with additions in line with what already done in #437.

**Comment by Ocramius** [comment]
Closing here: I had marked this as "won't fix", but didn't close 🤷
